### DISCUSSION TITLE
Template API: Exposing order and id to the client, for debugging/dev?

### DISCUSSION THREAD

**Discussion by mattsherman** [woocommerce/woocommerce#39968] [WooCommerce New Product Editor] [answered]
## Background

We've recently merged the core aspects of the product form template API, which will allow the modification (adding/removal) of blocks from the block-based product editor.

See:
- https://github.com/woocommerce/woocommerce/pull/39470
- https://github.com/woocommerce/woocommerce/pull/39900
- https://github.com/woocommerce/woocommerce/pull/39915
- https://github.com/woocommerce/woocommerce/pull/39946

(Documentation being worked on -- anticipate updates surrounding this in the coming weeks)

Blocks in the template are order based on an `order` property that is set on each block in the template. Extensions can use this to place their blocks in-between existing blocks, by setting the `order` appropriately.

Existing blocks in the template (such as a section that you want to place your block under) can be accessed by supplying the (template) `id` of the block.

## Issue

The `order` and `id` defined in the template are not sent to the client, so there is no way to look at the product form (or more accurately its rendered HTML source) and see what those are set to for a block in the form.

Instead, extension developers will need to look at the PHP source code of the template definition (our current, work-in-progress, default template is [SimpleProductTemplate](https://github.com/woocommerce/woocommerce/blob/a425a62fb8ec0749abc19cd69ebb6c811a1043c3/plugins/woocommerce/src/Admin/Features/ProductBlockEditor/ProductTemplates/SimpleProductTemplate.php)) and find the blocks they are interested in.

## Proposal

We should expose the `order` and `id` properties of blocks to the client in the rendered HTML source.

This would help extension developers more easily determine the values of those for blocks they are interested in.

It would also provide the ability for client-side tooling to be developed that exposed these properties in a more visual way (without having to inspect the HTML source).

## Implementation

Ideally, block developers would not need to do anything special to have these properties exposed to the client.

Is there a way in Gutenberg that we can attach arbitrary "meta-data" to blocks via a template to accomplish this? (/cc @joshuatf)

**Comment by louwie17** [comment]
> Is there a way in Gutenberg that we can attach arbitrary "meta-data" to blocks via a template to accomplish this?

There could be two way's, we could have our own wrapper of `useBlockProps`, as any props passed into that function get added to the block `div` -> `useBlockProps( { 'data-order': 5 } )` turns into `<div data-order="5" />`, but this is somewhat dependant on users making use of our block props function versus the main one.

There also appears to be a `getEditWrapperProps` function you can define in the block definition, so maybe we want to auto inject this with our template? See this example: https://github.com/WordPress/gutenberg/blob/4f7c19e37d55bb6a419cc8eca94c634f1e71852e/packages/block-library/src/text-columns/index.js#L16-L21

**Comment by joshuatf** [comment, marked as answer]
Originally I had planned to simply allow these properties to be passed via attributes `_templateBlockId` and `_templateOrder`.  I think this is still a viable and simple option.

However, if we want to keep these separate from `attributes` we could simply pass them in our soon-to-be block wrapper as a separate `template` object with template specific data.

```jsx
const templateProperties = [ '_blockId', '_order' ]
function WooEditWrapper( EditComponent ) {
    return ( { attributes } ) => {

        const template = templateProperties.reduce( ( props, property ) => {
            props[ property.slice(1) ] = attributes[ property ];
            return props;
        }, {} );

        const cleanedAttributes = { ...attributes };
        templateProperties.forEach( ( property ) => {
            cleanedAttributes.delete( property );
        } );

        return <EditComponent
            attributes={ cleanedAttributes }
            template={ template }
        />;
    }
}
```

**Comment by mattsherman** [comment]
Thanks, @louwie17 and @joshuatf -- I'll explore those approaches!
